core/src/main/java/kafka/server/SharePartitionManager.java

```diff
@@ -242,20 +242,24 @@ private ShareSessionKey shareSessionKey(String groupId, Uuid memberId) {
         return new ShareSessionKey(groupId, memberId);
     }
 
-    public Errors acknowledgeShareSessionCacheUpdate(String groupId, Uuid memberId, int reqEpoch) {
+    public Errors acknowledgeShareSessionCacheUpdate(String groupId, Uuid memberId, int reqEpoch, boolean isAcknowledgementPiggybackedOnFetch) {
         Errors shareAcknowledgeRequestError = shareAcknowledgeError(groupId, memberId, reqEpoch);
         if (shareAcknowledgeRequestError != Errors.NONE)
             return shareAcknowledgeRequestError;
         ShareSessionKey key = shareSessionKey(groupId, memberId);
         ShareSession shareSession = cache.get(new ShareSessionKey(groupId, memberId));
-        if (reqEpoch == ShareFetchMetadata.FINAL_EPOCH) {
-            if (cache.remove(key) != null)
-                log.info("Removed share session with key " + key);
-            return Errors.NONE;
+        // if acknowledgement is piggybacked on fetch, newContext function takes care of updating the share session epoch
+        // and share session cache. However, if the acknowledgement is standalone, the updates are handled in the if block
+        if (!isAcknowledgementPiggybackedOnFetch) {
+            if (reqEpoch == ShareFetchMetadata.FINAL_EPOCH) {
+                if (cache.remove(key) != null)
+                    log.info("Removed share session with key " + key);
+                return Errors.NONE;
+            }
+            // Update session's position in the cache
+            cache.touch(shareSession, time.milliseconds());
```

Comment:
I think you should touch the cache entry if there is a piggybacked ack to guard against the entry disappearing from the cache between the ack and the fetch.